extensions/amp-story/1.0/amp-story.js

```diff
@@ -953,6 +957,109 @@ export class AmpStory extends AMP.BaseElement {
   }
 
 
+  /**
+   * Switches to a particular page.
+   * Protected behing the 'amp-story-navigation-performance' experiment.
+   * @param {string} targetPageId
+   * @return {!Promise}
+   */
+  experimentalSwitchTo_(targetPageId) {
+    const targetPage = this.getPageById(targetPageId);
+    const pageIndex = this.getPageIndex(targetPage);
+
+    const oldPage = this.activePage_;
+    this.activePage_ = targetPage;
+
+    // Each step will run in a requestAnimationFrame, and wait for the next
+    // frame before executing the following step.
+    // First step contains the bare minimum the display and play the next page.
+    const steps = [
```

Comment:
I'm on the fence about what I'm about to say, so consider this optional: WDYT about making each step its own top-level function, for readability?

Reply:
I can think of pros for both implementations.

Three top level methods:
- Better naming (`willNavigate`, `navigate`, `didNavigate`?)
- Easier for a developer to know where to add their code (before, during, or after the navigation) without having to know about the "one step per frame" logic

All of these could be also achieved with better inline comments though.

Array of anonymous functions:
- Easier to see what's happening during the navigation without having to go through several methods (I personally think this version is easier to read)


I can see good things in both ideas, so really it's up to what you prefer

Reply:
SGTM to leave it as is, perhaps with a comment for each step